### PR TITLE
solution 1

### PR DESCRIPTION
**Pull request by kostya-kosyuk**
[DEMO LINK](https://kostya-kosyuk.github.io/react_add-todo-form/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,61 +1,129 @@
...
+  const handleAddTodo = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    const title = todoTitle.trim();
+
+    const id = visibleTodos
+      .reduce((maxId, todo) => (maxId > todo.id ? maxId : todo.id), 0) + 1;
+    const user = usersFromServer
+      .find(userFromServer => userFromServer.id === Number(choosedUserId)
+        ?? null);
```
Comment (andrii-dmytruk) [inline src/App.tsx:47]: You can use here func `getUser`

File: src/App.tsx [hunk 2]
```diff
@@ -1,61 +1,129 @@
...
+    if (title === '' || choosedUserId === '-1') {
+      setSubmit(true);
+    } else {
+      addVisibleTodo([...visibleTodos, {
+        id,
+        title,
+        userId,
+        completed,
+        user,
+      }]);
```
Comment (andrii-dmytruk) [inline src/App.tsx:60]: It's better to use fuctional version of `addVisibelTodo` if you need previouse value of state

**Review by andrii-dmytruk — CHANGES_REQUESTED** [review]
GJ! just little fixes

**Review by danylo-onishchuk — CHANGES_REQUESTED** [review]
test

**Review by andrii-dmytruk — APPROVED** [review]

**Review by Andr1yK — APPROVED** [review]
test

**Review by Andr1yK — APPROVED** [review]

**Review by Andr1yK — CHANGES_REQUESTED** [review]
test

File: src/App.tsx [hunk 3]
```diff
@@ -1,61 +1,127 @@
+import React, { useState } from 'react';
 import './App.scss';
 
-// import usersFromServer from './api/users';
-// import todosFromServer from './api/todos';
+import todosFromServer from './api/todos';
+import usersFromServer from './api/users';
+
+import { User } from './types/User';
```
Comment (AlinaMatuschak) [inline src/App.tsx:7]: Make sure that you pushed types files
Screenshot 2022-09-26 at 12 01 20

File: src/App.tsx [hunk 4]
```diff
@@ -1,61 +1,127 @@
...
+import { Todo } from './types/Todo';
+
+import { TodoList } from './components/TodoList';
+
+function getUser(userId: number): User | null {
+  const foundUser = usersFromServer.find(user => user.id === userId);
+
+  // if there is no user with a given userId
```
Comment (AlinaMatuschak) [inline src/App.tsx:15]: delete comment

File: src/App.tsx [hunk 5]
```diff
@@ -1,61 +1,127 @@
...
+  return foundUser || null;
+}
+
+export const todos: Todo[] = todosFromServer.map(todo => ({
+  ...todo,
+  user: getUser(todo.userId) ?? undefined,
```
Comment (AlinaMatuschak) [inline src/App.tsx:21]:
```suggestion
  user: getUser(todo.userId),
```

File: src/App.tsx [hunk 6]
```diff
@@ -1,61 +1,127 @@
...
+}));
+
+export const App: React.FC = () => {
+  const [choosedUserId, setChoosedUserId] = useState('-1');
```
Comment (AlinaMatuschak) [inline src/App.tsx:25]: better use null on default value

File: src/App.tsx [hunk 7]
```diff
@@ -1,61 +1,127 @@
...
+  };
+
+  const handleAddTodo = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    const title = todoTitle.trim();
+
+    const id = visibleTodos
+      .reduce((maxId, todo) => (maxId > todo.id ? maxId : todo.id), 0) + 1;
+    const user = getUser(Number(choosedUserId)) ?? undefined;
```
Comment (AlinaMatuschak) [inline src/App.tsx:45]:
```suggestion
    const user = getUser(Number(choosedUserId));
```

File: src/components/TodoInfo/TodoInfo.spec.tsx [hunk 8]
```diff
@@ -1,113 +0,0 @@
-import { mount } from '@cypress/react';
```
Comment (AlinaMatuschak) [inline src/components/TodoInfo/TodoInfo.spec.tsx:1]: why did you delete tests?)

**Review by AlinaMatuschak — CHANGES_REQUESTED** [review]
When I add spaces on the title, todo doesn’t create but error I don’t see too )
Screenshot 2022-09-26 at 12 05 04

**Review by andrii-dmytruk — APPROVED** [review]

File: src/App.tsx [hunk 9]
```diff
@@ -1,61 +1,128 @@
...
+
+  return foundUser || null;
+}
+
+export const todos: Todo[] = todosFromServer.map(todo => ({
+  ...todo,
+  user: getUser(todo.userId),
+}));
+
+const defaultId = '0';
```
Comment (AlinaMatuschak) [inline src/App.tsx:23]:
```suggestion
const DEFAULT_USER_ID = '0';
```

File: src/App.tsx [hunk 10]
```diff
@@ -1,61 +1,128 @@
...
+
+    const title = todoTitle.trim();
+
+    const id = visibleTodos
+      .reduce((maxId, todo) => (maxId > todo.id ? maxId : todo.id), 0) + 1;
+    const user = getUser(Number(choosedUserId));
+    const userId = user && user.id;
+    const completed = false;
+
+    if (title === '' || choosedUserId === '0' || !user || !userId) {
```
Comment (AlinaMatuschak) [inline src/App.tsx:50]: Why you don’t use the default variable?
```suggestion
    if (title === '' || choosedUserId === DEFAULT_USER_ID || !user || !userId) {
```

File: src/App.tsx [hunk 11]
```diff
@@ -1,61 +1,128 @@
...
+                <option
+                  value={user.id}
+                  key={user.id}
+                >
+                  {user.name}
+                </option>
+              ))}
+            </select>
+          </label>
+          {choosedUserId === '0'
```
Comment (AlinaMatuschak) [inline src/App.tsx:112]:
```suggestion
          {choosedUserId === DEFAULT_USER_ID
```

File: src/App.tsx [hunk 12]
```diff
@@ -1,61 +1,128 @@
...
+          <label>
+            {'User: '}
+            <select
+              data-cy="userSelect"
+              onChange={handleSelectingUser}
+              value={choosedUserId}
+            >
 
-          <span className="error">Please choose a user</span>
+              <option value="0" disabled>Choose a user</option>
```
Comment (AlinaMatuschak) [inline src/App.tsx:101]: DEFAULT_USER_ID 😂

**Review by AlinaMatuschak — APPROVED** [review]
